apps/src/lib/tools/jsinterpreter/JSInterpreter.js

```diff
@@ -0,0 +1,1096 @@
+        continue;
+      }
+      // If we've moved past the place of the last breakpoint hit without being
+      // deeper in the stack, we will discard the stoppedAtBreakpoint properties:
+      if (inUserCode && !this.findStoppedAtBreakpointRow(currentScope, userCodeRow)) {
+        this.removeStoppedAtBreakpointRowForScope(currentScope);
+      }
+      // If we're unwinding, continue to update the stoppedAtBreakpoint properties
+      // to ensure that we have the right properties stored when the unwind completes:
+      if (inUserCode && unwindingAfterStep) {
+        this.replaceStoppedAtBreakpointRowForScope(currentScope, userCodeRow);
+      }
+      if (this.logExecution) {
+        this.logStep_();
+      }
+      this.executionError = safeStepInterpreter(this);
+      if (!this.executionError && this.interpreter.getStackDepth()) {
+        const state = this.interpreter.peekStackFrame(), nodeType = state.node.type;
+
+        // Determine whether we are done executing a line of user code. This is detected by checking
+        // that one of the following conditions are true:
+        //   1. We've reached an "interstitial" node, which gets placed onto the stack
+        //      once an earlier expression has finished
+        //   2. The node on the stack has been explicitly marked as done by the interpreter
+        //   3. In the case of "update expression" nodes, like i++ or i--, the done state is tracked
+        //      in the doneLeft value, because the expression kind of looks like i = i + 1 which
+        //      has a left (i = ) and a right (i + 1) side, and the left side is where the assignment
+        //      actually takes place.
+        this.atInterstitialNode = INTERSTITIAL_NODES.hasOwnProperty(nodeType);
+        if (inUserCode && !doneUserLine) {
+          doneUserLine = (
+            this.atInterstitialNode ||
+            state.done ||
+            (state.node.type === 'UpdateExpression' && state.doneLeft)
+          );
+        }
+
+        const stackDepth = this.interpreter.getStackDepth();
+        // Remember the stack depths of call expressions (so we can implement 'step out')
+
+        // Truncate any history of call expressions seen deeper than our current stack position:
+        for (let depth = stackDepth + 1; depth <= this.maxValidCallExpressionDepth; depth++) {
+          this.callExpressionSeenAtDepth[depth] = false;
+        }
+        this.maxValidCallExpressionDepth = stackDepth;
+
+        if (inUserCode && this.interpreter.peekStackFrame().node.type === "CallExpression") {
+          // Store that we've seen a call expression at this depth in callExpressionSeenAtDepth:
+          this.callExpressionSeenAtDepth[stackDepth] = true;
+        }
+
+        if (this.paused) {
+          // Store the first call expression stack depth seen while in this step operation:
+          if (inUserCode && this.interpreter.peekStackFrame().node.type === "CallExpression") {
```

Comment:
I still think extracting `inUserCallExpression` (here and eight lines above) makes sense.